### PR TITLE
Changes for TripRequestService - service should have models

### PR DESCRIPTION
**Pull request by lakshmisuhas1996**

### CODE REVIEW

File: src/app/fyle/my-add-edit-trip/my-add-edit-trip.page.ts [hunk 1]
```diff
@@ -726,7 +727,7 @@ export class MyAddEditTripPage implements OnInit {
       this.tripRequest$ = this.tripRequestsService.get(id);
       const selectedProject$ = this.tripRequest$.pipe(
         switchMap(trip => {
-          return iif(() => trip.project_id, this.projectsService.getbyId(trip.project_id), of(null));
+          return iif(() => !!trip.project_id, this.projectsService.getbyId(trip.project_id), of(null));
```
Comment (madhavmansuriya40) [inline src/app/fyle/my-add-edit-trip/my-add-edit-trip.page.ts:730]: can you please add comments on why this change

File: src/app/fyle/view-team-trip/view-team-trip.page.ts [hunk 2]
```diff
@@ -32,9 +34,9 @@ export class ViewTeamTripPage implements OnInit {
 
   tripRequest$: Observable<ExtendedTripRequest>;
   approvals$: Observable<Approval[]>;
-  actions$: Observable<any>;
+  actions$: Observable<TripRequestActions>;
   actionsRedefined$: Observable<any>;
```
Comment (madhavmansuriya40) [inline src/app/fyle/view-team-trip/view-team-trip.page.ts:38]: if these are trip actions, any can be replaced with model

File: src/app/fyle/view-team-trip/view-team-trip.page.ts [hunk 3]
```diff
@@ -32,9 +34,9 @@ export class ViewTeamTripPage implements OnInit {
...
-  advanceRequests$: Observable<any>;
+  advanceRequests$: Observable<AdvanceRequest[]>;
   transportationRequests$: Observable<any>;
   hotelRequests$: Observable<any>;
```
Comment (madhavmansuriya40) [inline src/app/fyle/view-team-trip/view-team-trip.page.ts:41]: it will be good if we have hotel and, transport models in the same PR as they are part of trips

File: src/app/fyle/my-add-edit-trip/my-add-edit-trip.page.ts (hunk 1 above)
Comment (lakshmisuhas1996) [reply src/app/fyle/my-add-edit-trip/my-add-edit-trip.page.ts:730]: ts is complaining about boolean value is expected instead got string here... because of TripRequest model changes - project_id is string thus making this change to express it in boolean value

File: src/app/fyle/my-create-report/my-create-report.page.ts [hunk 4]
```diff
@@ -213,7 +214,7 @@ export class MyCreateReportPage implements OnInit {
           return (tripATime > tripBTime) ? -1 : ((tripATime < tripBTime) ? 1 : 0);
         });
       }),
-      map((tripRequests: any) => {
+      map((tripRequests) => {
         return tripRequests.map(tripRequest => {
           return {label: moment(tripRequest.created_at).format('MMM Do YYYY') + ', ' + tripRequest.purpose, value: tripRequest};
```
Comment (madhavmansuriya40) [inline src/app/fyle/my-create-report/my-create-report.page.ts:219]: `format('MMM Do YYYY')` can you please check this, I think this will break badly

File: src/app/fyle/my-edit-report/my-edit-report.page.ts [hunk 5]
```diff
@@ -270,7 +271,7 @@ export class MyEditReportPage implements OnInit {
           return (tripATime > tripBTime) ? -1 : ((tripATime < tripBTime) ? 1 : 0);
         });
       }),
-      map((tripRequests: any) => {
+      map((tripRequests) => {
         return tripRequests.map(tripRequest => {
           return {label: moment(tripRequest.created_at).format('MMM Do YYYY') + ', ' + tripRequest.purpose, value: tripRequest};
```
Comment (madhavmansuriya40) [inline src/app/fyle/my-edit-report/my-edit-report.page.ts:276]: `format('MMM Do YYYY')` same here, please check

File: src/app/fyle/my-edit-report/my-edit-report.page.ts [hunk 6]
```diff
@@ -280,7 +281,7 @@ export class MyEditReportPage implements OnInit {
 
   getSelectedTripInfo(tripRequestId) {
     if (this.canAssociateTripRequests) {
-      this.tripRequestsService.get(tripRequestId).subscribe((tripRequest: any) => {
+      this.tripRequestsService.get(tripRequestId).subscribe((tripRequest) => {
```
Comment (madhavmansuriya40) [inline src/app/fyle/my-edit-report/my-edit-report.page.ts:284]: can remove brackets here

File: src/app/core/models/custom_field.model.ts [hunk 7]
```diff
@@ -4,4 +4,8 @@ export interface CustomField {
     value: any;
     type?: string;
```
[the remark this answers is not in the document]
Comment (lakshmisuhas1996) [reply src/app/core/models/custom_field.model.ts:5]: Yes, custom fields are used in various places like in the response of advance_requests does not have all the properties and in hotel and transportation requests custom_field_values have all the props in the res...

File: src/app/core/models/hotel-request.model.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    trip_request_id: string;
+    trip_request_number: string;
+  };
+  requestor: {
+    department_id: string;
+  };
+  trp: {
+    project_id: string;
+  };
+  custom_field_values?: CustomField[];
```
[the remark this answers is not in the document]
Comment (lakshmisuhas1996) [reply src/app/core/models/hotel-request.model.ts:66]: in actual api response of Hotel Requests it will be not present. After unflattening the api res we are setting custom_field_values at some places

File: src/app/fyle/my-create-report/my-create-report.page.ts (hunk 4 above)
Comment (lakshmisuhas1996) [reply src/app/fyle/my-create-report/my-create-report.page.ts:219]: It is not breaking, if the created_at is 04-21-2021 and the value will become Apr 21st 2021

File: src/app/fyle/my-edit-report/my-edit-report.page.ts (hunk 5 above)
Comment (lakshmisuhas1996) [reply src/app/fyle/my-edit-report/my-edit-report.page.ts:276]: same here

**Review by madhavmansuriya40 — APPROVED** [review]

File: src/app/core/models/custom_field.model.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (iamtarun1993) [reply src/app/core/models/custom_field.model.ts:5]: If custom Fileds keys are different, then please create another Model for AdvanceRequest.

File: src/app/core/models/advance-request.model.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import { CustomField } from './custom_field.model';
```
Comment (iamtarun1993) [inline src/app/core/models/advance-request.model.ts:1]: move this file inside v1 / v2 directory

File: src/app/core/services/trip-requests.service.ts [hunk 10]
```diff
@@ -404,7 +409,7 @@ export class TripRequestsService {
   @CacheBuster({
     cacheBusterNotifier: tripRequestsCacheBuster$
   })
-  saveDraft(tripRequest) {
+  saveDraft(tripRequest: TripRequest) {
```
Comment (iamtarun1993) [inline src/app/core/services/trip-requests.service.ts:412]: output datatype of this method is not defined

File: src/app/core/services/trip-requests.service.ts [hunk 11]
```diff
@@ -418,7 +423,7 @@ export class TripRequestsService {
   @CacheBuster({
     cacheBusterNotifier: tripRequestsCacheBuster$
   })
-  submit(tripRequest) {
+  submit(tripRequest: TripRequest) {
```
Comment (iamtarun1993) [inline src/app/core/services/trip-requests.service.ts:426]: same as above

**Review by iamtarun1993 — APPROVED** [review]
